lms/static/scripts/frontend_apps/utils/format.js

```diff
@@ -0,0 +1,51 @@
+/**
+ * Truncate a URL to be `maxLength` or fewer characters.
+ *
+ * If `url` needs to be shortened, `truncateURL` tries to preserve the most
+ * informative parts of the URL using some simple heuristics.
+ *
+ * If `url` is not a valid URL, it is just shortened to the first `maxLength - 1`
+ * characters.
+ *
+ * @param {string} url
+ * @param {number} maxLength
+ */
+export function truncateURL(url, maxLength) {
+  if (url.length <= maxLength) {
+    return url;
+  }
+
+  // Strip the protocol and return if that is sufficient.
+  const urlWithoutScheme = url.replace(/^[^:]+:\/\//, '');
+  if (urlWithoutScheme.length <= maxLength) {
+    return urlWithoutScheme;
+  }
+
+  // Strip the query string and fragment, then continue removing path segments
+  // until the result is shorter than `maxLength`.
+  let parsed;
+  try {
+    parsed = new URL(url);
+  } catch {
+    return url.slice(0, maxLength - 1) + '…';
+  }
+
+  const hostname = parsed.hostname;
+  const pathSegments = parsed.pathname.split('/');
+  let pathTruncated = false;
+
+  const getCandidate = () =>
+    hostname + (pathTruncated ? '/…/' : '') + pathSegments.join('/');
+  while (pathSegments.length > 1 && getCandidate().length > maxLength) {
```

Comment:
I had to bench-check this with pen and paper to walk through it the first time. I think it would be helpful to add some commenting here. Something that explains what I believe to be the case: the while loop will chip away at the length of a long hostname-plus-path by removing one path segment at a time, from the left, until it either has a short enough string or runs out of path segments to remove. Not _complex_ but a comment would save a little future-time.

Reply:
I've replaced the descriptive comments with a worked example which I think makes the steps clearer.